src/server/handlers.rs

```diff
@@ -762,3 +747,158 @@ impl<P: Publisher> Handler<P> {
+    let schema_file = File::open("./oam/oam.schema.json")?;
+    Ok(BufReader::new(schema_file))
+}
+
+// Manifest validation
+pub(crate) fn validate_manifest(
+    manifest: Manifest,
+    current_manifests: StoredManifest,
+) -> Option<String> {
+    // let reader = open_schema_file().await.unwrap();
+    // let json_schema = serde_json::from_reader(reader).unwrap();
+    // let json_instance = serde_json::to_value(manifest.clone()).unwrap();
+    // let compiled_schema = JSONSchema::options()
+    //     .with_draft(Draft::Draft7)
+    //     .compile(&json_schema)
+    //     .expect("A valid schema");
+    // let validation_result = compiled_schema.validate(&json_instance);
+    // if let Err(errors) = validation_result {
+    //     for error in errors {
+    //         println!("Validation error: {}", error);
+    //         println!("Instance path: {}", error.instance_path);
+    //     }
+    //     return Some("Validation error".to_string());
+    // }
+
+    let latest_manifest = current_manifests.get_current();
+    let existing_linkdefs = latest_manifest.get_linkdef_map();
+
+    let mut component_details: HashSet<String> = HashSet::new();
+    let mut link_name_set: HashSet<String> = HashSet::new();
+
+    for component in manifest.spec.components.iter() {
+        // Component name validation : each component (actors or providers) should have a unique name
+        if !component_details.insert(component.name.clone()) {
+            return Some(format!(
+                "Duplicate component name in manifest: {}",
+                component.name
+            ));
+        }
+
+        // Provider validation :
+        // Provider config should be serializable [For all components that have JSON config, validate that it can serialize.
+        // We need this so it doesn't trigger an error when sending a command down the line]
+        // Providers should have a unique image ref and link name
+        if let Properties::Capability {
+            properties:
+                CapabilityProperties {
+                    image: image_name,
+                    link_name: Some(link),
+                    config: Some(CapabilityConfig::Json(data)),
+                    ..
+                },
+        } = &component.properties
+        {
+            if let Err(e) = serde_json::to_string(data) {
+                return Some(format!(
+                    "Unable to serialize JSON config data for component {}: {e:?}",
+                    component.name
+                ));
+            }
+
+            if !component_details.insert(image_name.to_string()) {
+                return Some(format!(
+                    "Duplicate image reference in manifest: {}",
+                    image_name
+                ));
+            }
+
+            if !link_name_set.insert(link.to_string()) {
+                return Some(format!("Duplicate link name in manifest: {}", image_name));
+            }
```

Comment:
This check actually needs to happen together for capability providers, e.g. the combination of the image name and link name needs to be unique, but if one of those differs then that's an issue.

Consider the following:
OK:
- image ref: httpserver:0.17.0, link name `default`
- image ref: httpserver:0.17.0, link name `notdefault`

OK:
- image ref: httpserver:0.17.0, link name `default`
- image ref: kvredis:0.17.0, link name `default`

Not OK:
- image ref: httpserver:0.17.0, link name `default`
- image ref: httpserver:0.17.0, link name `default`